controller/linodemachine_controller_helpers.go

```diff
@@ -107,20 +115,59 @@ func (r *LinodeMachineReconciler) newCreateConfig(ctx context.Context, machineSc
 	return createConfig, nil
 }
 
-func buildInstanceAddrs(linodeInstance *linodego.Instance) []clusterv1.MachineAddress {
+func (r *LinodeMachineReconciler) buildInstanceAddrs(ctx context.Context, logger logr.Logger, machineScope *scope.MachineScope, instanceID int) ([]clusterv1.MachineAddress, error) {
 	addrs := []clusterv1.MachineAddress{}
-	for _, addr := range linodeInstance.IPv4 {
-		addrType := clusterv1.MachineExternalIP
-		if addr.IsPrivate() {
-			addrType = clusterv1.MachineInternalIP
-		}
+	ips, err := r.getInstanceIPv4Addresses(ctx, logger, machineScope, instanceID)
+	if err != nil {
+		logger.Error(err, "Failed to get instance ip addresses")
+		return nil, err
+	}
+
+	// add all instance ips to machine's status
+	for _, ip := range ips {
 		addrs = append(addrs, clusterv1.MachineAddress{
-			Type:    addrType,
-			Address: addr.String(),
+			Type:    ip.ipType,
+			Address: ip.ip,
 		})
 	}
 
-	return addrs
+	return addrs, nil
+}
+
+func (r *LinodeMachineReconciler) getInstanceIPv4Addresses(ctx context.Context, logger logr.Logger, machineScope *scope.MachineScope, instanceID int) ([]nodeIP, error) {
+	addresses, err := machineScope.LinodeClient.GetInstanceIPAddresses(ctx, instanceID)
+	if err != nil {
+		return nil, err
+	}
+
+	// get the default instance config
+	configs, err := machineScope.LinodeClient.ListInstanceConfigs(ctx, instanceID, &linodego.ListOptions{})
+	if err != nil || len(configs) == 0 {
+		logger.Error(err, "Failed to list instance configs")
+		return nil, err
+	}
+
+	ips := []nodeIP{}
+	// check if a node has public ip and store it
+	if len(addresses.IPv4.Public) != 0 {
+		ips = append(ips, nodeIP{ip: addresses.IPv4.Public[0].Address, ipType: clusterv1.MachineExternalIP})
+	}
+
+	// Iterate over interfaces in config and find VPC specific ips
+	for _, iface := range configs[0].Interfaces {
+		if iface.VPCID != nil && iface.IPv4.VPC != "" {
+			ips = append(ips, nodeIP{ip: iface.IPv4.VPC, ipType: clusterv1.MachineInternalIP})
+		}
+	}
```

Comment:
Should we also filter these interfaces by the specific VPC referenced in the owner LinodeCluster's spec?

Reply:
I don't think we'll have machines connected to more than one VPC for now. If we do, we'll then need more than 2 interfaces and have to fix linode-ccm to figure out which interface is primary VPC interface etc etc. I feel current logic should be enough. Let me know if you think otherway and we can check further.

Reply:
Yeah, let's keep things simple for now then! 🙈

QQ: If users attempt to create a LinodeMachine with some interfaces already specified (e.g. via [`LinodeMachine.Spec.Interfaces`](https://github.com/linode/cluster-api-provider-linode/blob/db4b61cf1ec6a40512789689edc0c6f916aa9eaf/api/v1alpha1/linodemachine_types.go#L62C2-L62C12)), will that just break this feature?

This logic seems to _append_ to the `Interfaces` list in the `linodego.InstanceCreateOptions` and not overwrite it.

Reply:
TIL. Thank you for pointing to this. I tested and updated the logic to use machinetemplate to add additional interface.